keps/sig-cli/0031-kustomize-integration.md

```diff
@@ -0,0 +1,230 @@
+---
+kep-number: 31
+title: Enable kustomize in kubectl
+authors:
+  - "@Liujingfang1"
+owning-sig: sig-cli
+participating-sigs:
+  - sig-cli
+reviewers:
+  - "@pwittrock"
+  - "@seans3"
+  - "@soltysh"
+approvers:
+  - "@pwittrock"
+  - "@seans3"
+  - "@soltysh"
+editor: TBD
+creation-date: 2018-11-07
+last-updated: yyyy-mm-dd
+status: pending
+see-also:
+  - [KEP-0008](https://github.com/kubernetes/community/blob/master/keps/sig-cli/0008-kustomize.md)
+replaces:
+  - n/a
+superseded-by:
+  - n/a
+---
+
+# Enable kustomize in kubectl
+
+## Table of Contents
+* [Table of Contents](#table-of-contents)
+* [Summary](#summary)
+* [Motivation](#motivation)
+    * [Goals](#goals)
+    * [Non-Goals](#non-goals)
+* [Kustomize Introduction](#kustomize-introduction)    
+* [Proposal](#proposal)
+    * [UX](#UX)
+      * [apply](#apply)
+      * [get](#get)
+      * [delete](#delete)
+    * [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints)
+    * [Risks and Mitigations](#risks-and-mitigations)
+* [Graduation Criteria](#graduation-criteria)
+* [Implementation History](#implementation-history)
+* [Alternatives](#alternatives)
+
+[Tools for generating]: https://github.com/ekalinin/github-markdown-toc
+
+## Summary
+[Kustomize](https://github.com/kubernetes-sigs/kustomize) is a tool developed to provide declarative support for kubernetes objects. It has been adopted by many projects and users. Having kustomize enabled in kubectl will address a list of long standing issues. This KEP describes how `kustomize` is integrated into kubectl subcommands with consistent UX.
+
+## Motivation
+
+Declarative specification of Kubernetes objects is the recommended way to manage Kubernetes applications or workloads. There is some gap in kubectl on declarative support. To eliminate the gap, a [KEP](https://github.com/kubernetes/community/blob/master/keps/sig-cli/0008-kustomize.md#faq) was proposed months ago and Kustomize was developed. After more than 10 iterations, Kustomize has a complete set of features and reached a good state to be integrated into kubectl. 
+
+### Goals
+
+Integrate kustomize with kubectl so that kubectl can recognize kustomization directories and expand resources from kustomization.yaml before running kubectl subcommands. This integration should be transparent. It doesn't change kubectl UX. This integration should also be backward compatible. For non kustomization directories, kubectl behaves the same as current. The integration shouldn't have any impact on those parts.
```

Comment:
Does kustomize support capitalizing kustomization.yaml -- Kustomization.yaml? The reason Makefile and Dockerfile are capitalized is so they appear first in a directory listing, which is useful.

kubectl today doesn't infer any meaning from file names, and doesn't infer any meaning from file extensions, either, other than to identify files to process, if not explicitly specified files, URLs, or stdin. Instead, it requires object self-identification, which we do with all API types and componentconfig types.

I'm ok with starting with this as is, since it avoids the issues of what to do with multiple kustomization files and simplifies the config slightly, but we should be open to changing it later.

Reply:
It doesn't support capitalized file name currently. The filename `kustomization.yaml` is hard coded. We can change it to support different names later if desired. For example, in @pwittrock's kubectl book https://pwittrock-kubectl.firebaseapp.com/, the filename is `apply.yaml`

Reply:
Maybe it should be `Applyfile`?

Reply:
`apply.yaml` sounds better to me than `Applyfile`.